### PR TITLE
fix(pageserver): drain upload queue before detaching ancestor

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

In INC-317 https://neondb.slack.com/archives/C033RQ5SPDH/p1730815677932209, we saw an interesting series of operations that would remove valid layer files existing in the layer map.

* Timeline A starts compaction and generates an image layer Z but not uploading it yet.
* Timeline B/C starts ancestor detaching (which should not affect timeline A)
* The tenant gets restarted as part of the ancestor detaching process, without increasing the generation number.
* Timeline A reloads, discovering the layer Z is a future layer, and schedules a **deletion into the deletion queue**. This means that the file will be deleted any time in the future.
* Timeline A starts compaction and generates layer Z again, adding it to the layer map. Note that because we don't bump generation number during ancestor detach, it has the same filename + generation number as the original Z.
* Timeline A deletes layer Z from s3 + disk, and now we have a dangling reference in the layer map, blocking all compaction/logical_size_calculation process.

## Summary of changes

* We wait until all layers to be uploaded before shutting down the tenants in `Flush` mode.
* Ancestor detach restarts now use this mode.
* Ancestor detach also waits for remote queue completion before starting the detaching process.
* The patch ensures that we don't have any future image layer (or something similar) after restart, but not fixing the underlying problem around generation numbers.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by arpad-m** [comment]
It would be nice to have a test for this I think.

**Comment by skyzh** [comment]
@arpad-m I feel the exact scenario in the incident is hard to reproduce stably as it requires an ongoing but not finished compaction that creates an image layer to happen before detach ancestor. Therefore, I think asserting the queue is empty is a better way to do that -- unit tests will fail if the upload queue is not empty.

**Comment by arpad-m** [comment]
compaction on the same data should try to create the same files. But if it's too hard to create a regression test, let's just merge it as-is.

**Comment by skyzh** [comment]
simplified the patch and ready for review :)

File: pageserver/src/tenant/timeline.rs [hunk 1]
```diff
@@ -1747,6 +1747,14 @@ impl Timeline {
             }
         }
 
+        if try_freeze_and_flush {
```
Comment (jcsp) [inline pageserver/src/tenant/timeline.rs:1750]: isn't the previous block the same `if try freeze_and_flush` condition?  Can we move the code into that block?

Comment (skyzh) [reply pageserver/src/tenant/timeline.rs]: moved :) didn't notice that after many refactors of this patch... 🤣

**Review by arpad-m — APPROVED** [review]

**Comment by skyzh** [comment]
decided to add a flush-only mode to avoid flushing an extra L0 layer before shutdown

**Comment by skyzh** [comment]
would be good if @arpad-m would like to take another look at the newly added commit...

**Pull request merged**